packages/wonder-blocks-data/util/response-cache.js

```diff
@@ -0,0 +1,149 @@
+            _default = new ResponseCache();
+        }
+        return _default;
+    }
+
+    _cache: Cache = {};
+
+    constructor(source?: ?Cache = undefined) {
+        if (source != null) {
+            this.initialize(source);
+        }
+    }
+
+    _setCacheEntry<TOptions, TData>(
+        handler: IRequestHandler<TOptions, TData>,
+        options: TOptions,
+        entry: CacheEntry<TData>,
+    ): CacheEntry<TData> {
+        const requestType = handler.type;
+
+        // Ensure we have a cache location for this handler type.
+        this._cache[requestType] = this._cache[requestType] || {};
+
+        // Cache the data.
+        const key = handler.getKey(options);
+        this._cache[requestType][key] = Object.freeze(entry);
+        return this._cache[requestType][key];
+    }
+
+    /**
+     * Initialize the cache from a given cache state.
+     *
+     * This can only be called if the cache is not already in use.
+     */
+    initialize = (source: Cache): void => {
+        if (Object.keys(this._cache).length !== 0) {
+            throw new Error(
+                "Cannot initialize data response cache more than once",
+            );
+        }
+
+        try {
+            /**
+             * Object.assign only performs a shallow clone.
+             * So we deep clone it and then assign the clone values to our
+             * internal cache.
+             */
+            const cloneInitCache = deepClone(source);
+            Object.assign(this._cache, cloneInitCache);
+        } catch (e) {
+            throw new Error(
+                `An error occurred trying to initialize the data response cache: ${e}`,
+            );
+        }
+    };
+
+    /**
+     * Cache data for a specific response.
+     */
+    cacheData = <TOptions, TData>(
+        handler: IRequestHandler<TOptions, TData>,
+        options: TOptions,
+        data: TData,
+    ): CacheEntry<TData> => {
+        return this._setCacheEntry(handler, options, {data});
+    };
+
+    /**
+     * Cache an error for a specific response.
+     */
+    cacheError = <TOptions, TData>(
+        handler: IRequestHandler<TOptions, TData>,
+        options: TOptions,
+        error: Error | string,
+    ): CacheEntry<TData> => {
+        const errorMessage = typeof error === "string" ? error : error.message;
+        return this._setCacheEntry(handler, options, {error: errorMessage});
+    };
+
+    /**
+     * Retrieve data from our cache.
+     */
+    getEntry = <TOptions, TData>(
+        handler: IRequestHandler<TOptions, TData>,
+        options: TOptions,
+    ): ?$ReadOnly<CacheEntry<TData>> => {
+        const requestType = handler.type;
+
+        // Get the subcache for the handler.
+        const handlerCache = this._cache[requestType];
+        if (!handlerCache) {
+            return null;
+        }
+
+        const key = handler.getKey(options);
+        if (handler.invalidateCache(options)) {
+            delete handlerCache[key];
+        }
```

Comment:
When would we want `getEntry()` to invalidate the cache?

Reply:
When a handler says it should. For example, perhaps it's been 1 day since the page was loaded and we do a client-side nav. Maybe for a given piece of data, that means the handler wants to re-request. This gives us the ability to do that.